File: ios/Sources/FindUsEngine/IncidentSession.swift
import Foundation
import CryptoKit

/// Incident session (port of core/security.py): double-HMAC key derivation —
/// byte-for-byte identical to Python/Kotlin so one link authenticates on every OS.
public final class IncidentSession {
    public static let macLength = 8
    public static let linkPrefix = "incident://"
    private static let incidentFixedKey = "CROWD-COMPASS/INCIDENT"
    private static let hkdfInfo = "crowd-compass/incident-session-v1"
    private static let authInfo = "crowd-compass/gradient-mac-v1"
    private static let pseudonymInfo = "crowd-compass/incident-pseudonym-v1"

    public let incidentId: String
    public let salt: Data
    public var linkTtl: Int
    public var linkLife: Double
    public let key: Data
    public let authKey: Data

    public init(incidentId: String, salt: Data, linkTtl: Int = 8, linkLife: Double = 900.0) {
        self.incidentId = incidentId
        self.salt = salt
        self.linkTtl = linkTtl
        self.linkLife = linkLife
        let msg = Data(incidentId.utf8) + Data([0]) + salt
        let prk = Self.hmac(key: Data(Self.incidentFixedKey.utf8), data: msg)
        self.key = Self.hmac(key: prk, data: Data(Self.hkdfInfo.utf8))
        self.authKey = Self.hmac(key: self.key, data: Data(Self.authInfo.utf8))
    }

    private static func hmac(key: Data, data: Data) -> Data {
        let mac = HMAC<SHA256>.authenticationCode(for: data, using: SymmetricKey(data: key))
        return Data(mac)
    }

    public func mac(_ pkt: RelayPacket) -> Data {
        let full = Self.hmac(key: authKey, data: canonicalFields(pkt))
        return full.prefix(Self.macLength)
    }

    public func verify(_ pkt: RelayPacket) -> Bool {
        guard let tag = pkt.authTag else { return false }
        return mac(pkt) == tag.prefix(Self.macLength)
    }

    public func attach(_ pkt: RelayPacket) -> RelayPacket {
        var signed = pkt
        signed.authTag = mac(pkt)
        return signed
    }

    public func pseudonym(installKey: Data) -> String {
        let raw = Self.hmac(key: installKey,
                            data: Data(Self.pseudonymInfo.utf8) + Data([0]) + Data(incidentId.utf8))
        return base64URL(raw.prefix(6))
    }

    public func toLink(ttlHops: Int = 8, lifespan: Double = 900.0) -> String {
        let idData = Data(incidentId.utf8)
        var payload = Data()
        UInt16(idData.count).bigEndian.withUnsafeBytes { payload.append(contentsOf: $0) }
        payload.append(idData)
        payload.append(salt)
        let saltB64 = base64URL(salt)
        return "\(Self.linkPrefix)\(base64URL(payload))?salt=\(saltB64)&ttl=\(ttlHops)&life=\(Int(lifespan))"
    }

    public static func makeSession(incidentId: String, salt: Data = randomSalt()) -> IncidentSession {
        IncidentSession(incidentId: incidentId, salt: salt)
    }

    public static func randomSalt(size: Int = 16) -> Data {
        var bytes = [UInt8](repeating: 0, count: size)
        _ = SecRandomCopyBytes(kSecRandomDefault, size, &bytes)
        return Data(bytes)
    }

    public static func parseLink(_ link: String) throws -> IncidentSession {
        guard link.hasPrefix(linkPrefix) else {
            throw LinkError.invalid("not an incident link")
        }
        let rest = String(link.dropFirst(linkPrefix.count))
        let query = rest.split(separator: "?", maxSplits: 1)
        let body = String(query.first ?? "")
        let params = query.count > 1 ? parseParams(String(query[1])) : [:]
        let decoded = try base64URLData(body)
        guard decoded.count >= 3 else {
            throw LinkError.invalid("malformed link payload")
        }
        guard decoded.count >= 2 else {
            throw LinkError.invalid("malformed link payload")
        }
        let header = Array(decoded.prefix(2))
        let idLen = Int(UInt16(header[0]) << 8 | UInt16(header[1]))
        guard 2 + idLen <= decoded.count else {
            throw LinkError.invalid("malformed link payload")
        }
        guard let incidentId = String(data: decoded.dropFirst(2).prefix(idLen), encoding: .utf8) else {
            throw LinkError.invalid("bad incident id encoding")
        }
        let salt = decoded.dropFirst(2 + idLen)
        let session = IncidentSession(incidentId: incidentId, salt: Data(salt))
        session.linkTtl = Int(params["ttl"] ?? "") ?? 8
        session.linkLife = Double(params["life"] ?? "") ?? 900.0
        return session
    }

    /// Canonical on-wire fields: U8 type | 16B event | 8B sender | U16 ver | U8 hop | U8 ttl | U32 ts.
    public func canonicalFields(_ pkt: RelayPacket) -> Data {
        var out = Data()
        out.append(UInt8(pkt.messageType.rawValue))
        out.append(Self.padID(pkt.eventId?.value ?? "", width: 16))
        out.append(Self.padID(pkt.sender?.value ?? "", width: 8))
        out.append(UInt8((pkt.sourceVersion >> 8) & 0xFF))
        out.append(UInt8(pkt.sourceVersion & 0xFF))
        out.append(UInt8(min(max(pkt.hop, 0), 0xFF)))
        out.append(UInt8(min(max(pkt.ttl, 0), 0xFF)))
        out.append(canonTS(pkt.timestamp))
        return out
    }

    private func canonTS(_ t: Double) -> Data {
        guard t != EngineConstants.timeNone, t > 0 else {
            return Data([0, 0, 0, 0])
        }
        let v = UInt32(clamping: Int64((t - EngineConstants.epoch).rounded()))
        return Data([UInt8((v >> 24) & 0xFF), UInt8((v >> 16) & 0xFF),
                     UInt8((v >> 8) & 0xFF), UInt8(v & 0xFF)])
    }

    private static func padID(_ value: String, width: Int) -> Data {
        var out = Data(repeating: 0, count: width)
        let raw = Data(value.utf8)
        for i in 0..<min(raw.count, width) {
            out[out.startIndex + i] = raw[raw.index(raw.startIndex, offsetBy: i)]
        }
        return out
    }

    public enum LinkError: Error {
        case invalid(String)
    }
}

// base64url helpers (shared with the app layer)
public func base64URL(_ data: Data) -> String {
    data.base64EncodedString()
        .replacingOccurrences(of: "+", with: "-")
        .replacingOccurrences(of: "/", with: "_")
        .replacingOccurrences(of: "=", with: "")
}

public func base64URLData(_ s: String) throws -> Data {
    var v = s
        .replacingOccurrences(of: "-", with: "+")
        .replacingOccurrences(of: "_", with: "/")
    while v.count % 4 != 0 { v.append("=") }
    guard let d = Data(base64Encoded: v) else {
        throw IncidentSession.LinkError.invalid("bad base64")
    }
    return d
}

private func parseParams(_ query: String) -> [String: String] {
    var out: [String: String] = [:]
    for part in query.split(separator: "&") {
        let kv = part.split(separator: "=", maxSplits: 1)
        if kv.count == 2 { out[String(kv[0])] = String(kv[1]) }
    }
    return out
}

File: android/app/src/main/kotlin/com/findus/crowdcompass/engine/IncidentSession.kt
package com.findus.crowdcompass.engine

import java.security.MessageDigest
import java.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Incident session port of core/security.py: shared key derived from the
 * incident link, 8-byte canonical HMAC-SHA256, incident-scoped pseudonym.
 */
class IncidentSession(val incidentId: String, val salt: ByteArray,
                      var linkTtl: Int = 8, var linkLife: Double = 900.0) {

    val key: ByteArray = hmac("HmacSHA256", hmac("HmacSHA256",
        INCIDENT_FIXED_KEY.toByteArray(), msgBytes()), HKDF_INFO)
    val authKey: ByteArray = hmac("HmacSHA256", key, AUTH_INFO)

    private fun msgBytes(): ByteArray {
        val out = ByteArrayOutputStreamSafe()
        out.write(incidentId.toByteArray(Charsets.UTF_8))
        out.write(0)
        out.write(salt)
        return out.toByteArray()
    }

    fun mac(pkt: RelayPacket): ByteArray {
        val full = hmac("HmacSHA256", authKey, canonicalFields(pkt))
        return full.copyOf(MAC_LENGTH)
    }

    fun verify(pkt: RelayPacket): Boolean {
        pkt.authTag ?: return false
        val expected = mac(pkt)
        return MessageDigest.isEqual(expected, pkt.authTag.copyOf(MAC_LENGTH))
    }

    fun attach(pkt: RelayPacket): RelayPacket =
        pkt.copy(authTag = mac(pkt))

    fun pseudonym(installKey: ByteArray): String {
        val raw = hmac("HmacSHA256", installKey,
            PSEUDOOM_INFO + byteArrayOf(0) + incidentId.toByteArray(Charsets.UTF_8))
        return Base64.getUrlEncoder().withoutPadding().encodeToString(raw.copyOf(6))
    }

    fun toLink(ttlHops: Int = 8, lifespan: Double = 900.0): String {
        val saltB64 = Base64.getUrlEncoder().withoutPadding().encodeToString(salt)
        val idBytes = incidentId.toByteArray(Charsets.UTF_8)
        val payloadBytes = ByteArray(2 + idBytes.size + salt.size)
        payloadBytes[0] = (idBytes.size ushr 8).toByte()
        payloadBytes[1] = idBytes.size.toByte()
        idBytes.copyInto(payloadBytes, 2)
        salt.copyInto(payloadBytes, 2 + idBytes.size)
        val payload = Base64.getUrlEncoder().withoutPadding().encodeToString(payloadBytes)
        return "$LINK_PREFIX$payload?salt=$saltB64&ttl=$ttlHops&life=${lifespan.toInt()}"
    }

    companion object {
        const val MAC_LENGTH = 8
        const val LINK_PREFIX = "incident://"
        const val INCIDENT_FIXED_KEY = "CROWD-COMPASS/INCIDENT"
        val HKDF_INFO = "crowd-compass/incident-session-v1".toByteArray()
        val AUTH_INFO = "crowd-compass/gradient-mac-v1".toByteArray()
        val PSEUDOOM_INFO = "crowd-compass/incident-pseudonym-v1".toByteArray()

        fun makeSession(incidentId: String, salt: ByteArray = randomSalt()): IncidentSession =
            IncidentSession(incidentId, salt)

        fun randomSalt(): ByteArray {
            val rng = java.security.SecureRandom()
            return ByteArray(16).also { rng.nextBytes(it) }
        }

        fun parseLink(link: String): IncidentSession {
            if (!link.startsWith(LINK_PREFIX)) {
                throw IllegalArgumentException("not an incident link")
            }
            val rest = link.removePrefix(LINK_PREFIX)
            val query = rest.substringAfter('?', "")
            val bodyB64 = rest.substringBefore('?')
            val padded = bodyB64 + "=".repeat((4 - bodyB64.length % 4) % 4)
            val decoded = Base64.getUrlDecoder().decode(padded)
            if (decoded.size < 3) throw IllegalArgumentException("malformed link payload")
            val n = ((decoded[0].toInt() and 0xFF) shl 8) or (decoded[1].toInt() and 0xFF)
            if (2 + n > decoded.size) throw IllegalArgumentException("malformed link payload")
            val incidentId = String(decoded, 2, n, Charsets.UTF_8)
            val salt = decoded.copyOfRange(2 + n, decoded.size)
            val session = IncidentSession(incidentId, salt)
            session.linkTtl = queryParam(query, "ttl", 8) { it.toIntOrNull() ?: 8 }
            session.linkLife = queryParam(query, "life", 900.0) { it.toDoubleOrNull() ?: 900.0 }
            return session
        }

        private inline fun <T> queryParam(query: String, key: String, default: T, conv: (String) -> T): T {
            for (part in query.split("&")) {
                val kv = part.split("=", limit = 2)
                if (kv.size == 2 && kv[0] == key) return conv(kv[1])
            }
            return default
        }

        private fun hmac(algo: String, key: ByteArray, data: ByteArray): ByteArray {
            val m = Mac.getInstance(algo)
            m.init(SecretKeySpec(key, algo))
            return m.doFinal(data)
        }
    }

    /** Canonical on-wire fields: U8 mtype | 16B event | 8B sender | U16 ver | U8 hop | U8 ttl | U32 ts. */
    fun canonicalFields(pkt: RelayPacket): ByteArray {
        val ts = canonTs(pkt.timestamp)
        return ByteBufferBE()
            .put(pkt.messageType.wire.toByte())
            .put(padId(pkt.eventId?.value.orEmpty(), 16))
            .put(padId(pkt.sender?.value.orEmpty(), 8))
            .putShort(pkt.sourceVersion.coerceIn(0, 0xFFFF).toShort())
            .put(pkt.hop.coerceIn(0, 0xFF).toByte())
            .put(pkt.ttl.coerceIn(0, 0xFF).toByte())
            .putInt(ts)
            .array()
    }

    private fun canonTs(t: Double): Int {
        if (t == Engine.TIME_NONE || t <= 0.0) return 0
        return Math.round(t - Engine.EPOCH).coerceIn(0, 0xFFFFFFFFL).toInt()
    }

    private fun padId(value: String, width: Int): ByteArray {
        val raw = value.toByteArray(Charsets.UTF_8)
        val out = ByteArray(width)
        for (i in 0 until minOf(raw.size, width)) out[i] = raw[i]
        return out
    }
}

/** Minimal big-endian byte writer to keep protocol + MAC byte-faithful. */
class ByteBufferBE {
    private val data = java.io.ByteArrayOutputStream()
    fun put(b: Byte): ByteBufferBE { data.write(b.toInt()); return this }
    fun put(bs: ByteArray): ByteBufferBE { data.write(bs); return this }
    fun putShort(v: Short): ByteBufferBE {
        data.write((v.toInt() ushr 8) and 0xFF); data.write(v.toInt() and 0xFF); return this
    }
    fun putInt(v: Int): ByteBufferBE {
        data.write((v ushr 24) and 0xFF); data.write((v ushr 16) and 0xFF)
        data.write((v ushr 8) and 0xFF); data.write(v and 0xFF); return this
    }
    fun array(): ByteArray = data.toByteArray()
}

class ByteArrayOutputStreamSafe {
    private val data = java.io.ByteArrayOutputStream()
    fun write(b: Int) { data.write(b) }
    fun write(bs: ByteArray) { data.write(bs) }
    fun toByteArray(): ByteArray = data.toByteArray()
}